lucene/highlighter/src/java/org/apache/lucene/search/matchhighlight/MatchRegionRetriever.java

```diff
@@ -0,0 +1,503 @@
+                          + fieldInfo.getIndexOptions());
+                };
+        }
+        offsetStrategies.put(field, offsetStrategy);
+      }
+    }
+    return offsetStrategies;
+  }
+
+  public interface HitRegionConsumer {
+    void accept(LeafReader leafReader, int leafDocId, Map<String, List<OffsetRange>> hits)
+        throws IOException;
+  }
+
+  /**
+   * An abstraction that provides document values for a given field. Default implementation
+   * in {@link DocumentFieldValueProvider} just reaches to a preloaded {@link Document}. It is
+   * possible to write a more efficient implementation on top of a reusable character buffer
+   * (that reuses the buffer while retrieving hit regions for documents).
+   */
+  public interface FieldValueProvider {
+    List<CharSequence> getValues(String field);
+  }
+
+  public static final class DocumentFieldValueProvider implements FieldValueProvider {
+    private final Document doc;
+
+    public DocumentFieldValueProvider(Document doc) {
+      this.doc = doc;
+    }
+
+    @Override
+    public List<CharSequence> getValues(String field) {
+      return Arrays.asList(doc.getValues(field));
+    }
+  }
+
+  /**
+   * Determines how match offset regions are computed from {@link MatchesIterator}. Several
+   * possibilities exist, ranging from retrieving offsets directly from a match instance
+   * to re-evaluating the document's field and recomputing offsets from there.
+   */
+  private interface OffsetsFromMatchesStrategy {
+    List<OffsetRange> get(MatchesIterator matchesIterator, FieldValueProvider doc)
+        throws IOException;
+
+    default boolean requiresDocument() {
+      return false;
+    }
+  }
+
+  /**
+   * This strategy retrieves offsets directly from {@link MatchesIterator}.
+   */
+  private static class OffsetsFromMatchIterator implements OffsetsFromMatchesStrategy {
+    private final String field;
+
+    OffsetsFromMatchIterator(String field) {
+      this.field = field;
+    }
+
+    @Override
+    public List<OffsetRange> get(MatchesIterator matchesIterator, FieldValueProvider doc)
+        throws IOException {
+      ArrayList<OffsetRange> ranges = new ArrayList<>();
+      while (matchesIterator.next()) {
+        int from = matchesIterator.startOffset();
+        int to = matchesIterator.endOffset();
+        if (from < 0 || to < 0) {
+          throw new IOException("Matches API returned negative offsets for field: " + field);
+        }
+        ranges.add(new OffsetRange(from, to));
+      }
+      return ranges;
+    }
+  }
+
+  /**
+   * This strategy works for fields where we know the match occurred but there are
+   * no known positions or offsets.
+   *
+   * We re-analyze field values and return offset ranges for entire values
+   * (not individual tokens). Re-analysis is required because analyzer may return
+   * an unknown offset gap.
+   */
+  private static class OffsetsFromValues implements OffsetsFromMatchesStrategy {
```

Comment:
Right. I experimented with both. Don't know if it should be configurable or somehow tweakable. Perhaps. Maybe the strategy-creating method could be protected or somehow exposed so that people can choose. Like I said in another thread, I wouldn't want it to become too bloated.